[Feat] #140 - Push Notification 구현

joho2022 opened the pull request:
## 🌴 작업한 브랜치
- #140 


## ✅ 작업한 내용
<!-- 작업한 내용 적어주세요! -->

- [x] 푸터뷰 구현완료
- [x] 알림권한 모달 및 설정 구현완료
- [x] 푸시 알림 구현완료
- [ ] 딥링크 구현

<!--
```
넣고싶은 코드가 있다면 적어주세요
```
-->


## ❗️PR Point
<!-- 부족했던 점 혹은 개선하고 싶은 방향이 있다면 얘기해주세요 -->
채팅탭 진입 시
case os권한 on, 앱내 권한 off
case os권한 off, 앱내 권한 on
case os권한 off, 앱내 권한 off

케이스가 해당되면 최초 1회 모달 표시가 되고,
이후 각 권한 on이 되고 나서 다시 off가 된다면 다시 최초 1회 모달표시가 나타나게 됩니다.


<!--
```
넣고싶은 코드가 있다면 적어주세요
```
-->


## 📸 스크린샷
<!-- 스크린 샷을 첨부해주세요 -->
|기능|스크린샷|
|:--:|:--:|
|모달표시|<img src = "https://github.com/user-attachments/assets/9d7c6b3c-c69a-4610-bba2-37a881b18c28" width ="250">|
|알림 수신|<img src = "https://github.com/user-attachments/assets/16dddf9a-6d43-4da3-b81a-7e8015e40ec5" width ="250">|




## 📟 관련 이슈
- Resolved: #140

joho2022 committed:
[Feat] #140 - AppDelegate 초기 설정 및 PushSettingManager 구현

joho2022 committed:
[Chore] #140 - 싱글톤 제거 및 AuthManager 인스턴스화

joho2022 committed:
[Network] #140 - Push Network관련 DTO, API, Service 구현

joho2022 committed:
[Chore] #140 - APNs 개발 환경 설정 추가

joho2022 committed:
[Add] #140 - 이미지 에셋 추가

joho2022 committed:
[Chore] #140 - Firebase 설정 파일 .gitignore에 추가

joho2022 committed:
[Style] #140 - FooterView 구현

joho2022 committed:
[Add] #140 - 이미지 에셋 추가

joho2022 committed:
[Style] #140 - PermissionModalView 구현

joho2022 committed:
[Add] #140 - NetworkService에 PushService 추가

joho2022 committed:
[Feat] #140 - 푸시 권한/토큰 관리 구조 개선 및 PushManager 추가

joho2022 committed:
[Chore] #140 - 푸시 매니저 DI 리팩토링 및 초기화 개선

joho2022 committed:
[Merge] #140 - pull develop

joho2022 committed:
[Feat] #140 - 텍스트 복사기능 추가

joho2022 committed:
[Feat] #140 - CustomToggleStyle 추가

joho2022 committed:
[Feat] #140 - View 중앙 배치 centerInParent Modifier 추가

joho2022 committed:
[Feat] #140 - body7Medium16 폰트 스타일 추가

joho2022 committed:
[Feat] #140 - 푸시 권한 및 모달 로직 통합 개선

joho2022 committed:
[Merge] #140 - pull develop

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Components/Modal/View/PermissionModalView.swift:
```diff
@@ -0,0 +1,96 @@
+
+import SwiftUI
+
+struct PermissionModalView: View {
+    
```
P2

Modal 보다는 Alert의 느낌이 더 강한 것 같은데 어떻게 생각하시는지요!

joho2022 replied:
피그마에서 Modal로 표현되어 있어서 Modal로 네이밍했는데,
Alert이 더 적합해보이네요!

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Push/PushManager.swift:
```diff
@@ -0,0 +1,124 @@
+//  Created by 조호근 on 7/8/25.
+//
+
+import UIKit
```
P2

안 쓰는 프레임 워크는 제거해주세여

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Push/PushManager.swift:
```diff
@@ -0,0 +1,124 @@
+    ) async -> UNNotificationPresentationOptions {
+        
+        guard permission.shouldShowPush else {
+            print("앱/OS 알림꺼짐 -> 알림 무시")
+            return []
+        }
+        
+        // TODO: - 채팅방일 때 알림 안 받도록
+        
+        return [.banner, .list, .sound, .badge]
+    }
+
+    func userNotificationCenter(
+        _ center: UNUserNotificationCenter,
+        didReceive response: UNNotificationResponse
+    ) async {
+        // TODO: - deep link 처리 or Chat 이동 로직
+        
+        let userInfo = response.notification.request.content.userInfo
+        print("받은 푸시 데이터: ", userInfo)
+        print("알림 제목: ", response.notification.request.content.title)
```
P2

print -> logger 😎

joho2022 replied:
데이터 어떻게 들어오는지 구현시에 확인할려 했어서, 딥링크 구현시에 지울게요!

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Push/PushPermissionManager.swift:
```diff
@@ -0,0 +1,83 @@
+        isAppPushEnabled && isOSPushEnabled
+    }
+    
+    var pushOffState: PushOffState? {
+        switch (isAppPushEnabled, isOSPushEnabled) {
+        case (false, true):
+            return .appOnlyOff
+        case (true, false):
+            return .osOnlyOff
+        case (false, false):
+            return .bothOff
+        case (true, true):
+            return nil
+        }
```
분기 처리 넘 깔끔 꿋

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Extensions/View+.swift:
```diff
@@ -28,4 +28,11 @@ extension View {
         UIApplication.shared
             .sendAction(#selector(UIResponder.resignFirstResponder), to: nil, from: nil, for: nil)
     }
+    
+    func centerInParent() -> some View {
+        GeometryReader { geometry in
+            self
+                .position(x: geometry.size.width / 2, y: geometry.size.height / 2)
+        }
+    }
```
오 요건 뭔가요??

joho2022 replied:
알럿창 띄울 때 쓸려고 했는데, 재사용 가능성 있다고 생각해서 해당 메서드 호출하면 부모뷰기준으로 정중앙에 위치합니다!

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Style/CustomToggleStyle.swift:
```diff
@@ -0,0 +1,53 @@
+
+import SwiftUI
+
+struct CustomToggleStyle: ToggleStyle {
+    func makeBody(configuration: Configuration) -> some View {
```
오 첨 알아갑니다 굿굿

Johyerin reviewed (approved):
LGTM
푸시 수신, 알림 설정 창 케이스에 따라 잘 동작하는 것 확인 완료요
코드 깔끔하고 좋으네요

Johyerin commented on Napzak-iOS/Napzak-iOS/Presentation/Home/View/HomeView.swift:
```diff
@@ -94,6 +92,12 @@ struct HomeView: View {
             }
         }
+        .onAppear {
+          Task {
+            await pushManager.configureNotifications()
+            await pushManager.upsertTokenIfNeeded()
+          }
```
P1

들여쓰기가 요상합니다요

Johyerin reviewed (approved):
하나만 추가..

OneTen19 commented on Napzak-iOS/Napzak-iOS/Application/AppDelegate.swift:
```diff
@@ -0,0 +1,28 @@
+import UserNotifications
+import FirebaseMessaging
+
+class AppDelegate: NSObject, UIApplicationDelegate {
```
UIKit 의 AppDelegate랑 같은건가요?? 
스유 프로젝트에는 기본적으로 없길래 못 쓰는 줄 알았는데 그냥 이렇게 만들어서 쓰면 되나 보군여

joho2022 replied:
네 동일합니다!

OneTen19 commented on Napzak-iOS/Napzak-iOS/Application/Napzak_iOSApp.swift:
```diff
@@ -8,15 +8,29 @@
 
 @main
 struct Napzak_iOSApp: App {
+    @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate
```
이렇게 쓰는군

OneTen19 commented on Napzak-iOS/Napzak-iOS/Global/Components/Modal/View/PermissionModalView.swift:
```diff
@@ -0,0 +1,96 @@
+                    Button {
+                        print("close")
+                        onDismiss?()
+                        
```
p3

요기 공백!

OneTen19 commented on Napzak-iOS/Napzak-iOS/Global/Core/Push/PushManager.swift:
```diff
@@ -0,0 +1,124 @@
+//
+//  PushManager.swift
```
지금 프로젝트 전체적으로 매니저들이 좀 많이 생긴 것 같은데,
얘네들도 Manager 폴더로 따로 묶는 게 보기 편하려나요...

joho2022 replied:
저는 개인적으로 Manager들을 따로 폴더링하면 오히려 파악하는 데 더 피곤할 수 있다고 생각해요!
기능별로 매니저가 흩어져 있어도, 각 역할이 명확히 드러난다면 그게 더 직관적이라고 생각합니다.
기능 흐름 따라 코드 추적할 때도, 기능 단위로 폴더링되어있으면 이해하기 쉬울 것 같아요

OneTen19 commented on Napzak-iOS/Napzak-iOS/Global/Core/Push/PushManager.swift:
```diff
@@ -0,0 +1,124 @@
+        guard await permission.requestNotificationPermission() else { return }
+    }
+    
+    func upsertTokenIfNeeded() async {
```
upsert가 update랑 insert 합친거구나... 처음 알았다

lalaurrel replied:
오 신기

OneTen19 reviewed (approved):
수고하셨습니다👍

lalaurrel reviewed (approved):
고생하셨습니다람쥐

joho2022 committed:
[Fix] #140 - 코드리뷰 반영

joho2022 committed:
[Fix] #140 - 코드리뷰 피드백 반영

joho2022 committed:
[Merge] #140 - pull develop

# Conflicts:
#	Napzak-iOS/Napzak-iOS.xcodeproj/project.pbxproj
#	Napzak-iOS/Napzak-iOS/Global/Core/Auth/AuthManager.swift
#	Napzak-iOS/Napzak-iOS/Network/Base/NetworkService.swift
#	Napzak-iOS/Napzak-iOS/Presentation/TabBar/View/NZTabBarView.swift
